# Oakland 311 calls

This notebook is part of Assignment #6.

In [1]:
import pandas as pd
import altair as alt
import requests

/Users/gracelynnewest/.pyenv/versions/3.8.5/envs/lecture_06-3.8.5/lib/python3.8/site-packages/pandas/compat/__init__.py:124: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.
  warnings.warn(msg)


## Download data

Download [311 call data](https://data.oaklandca.gov/Infrastructure/Service-requests-received-by-the-Oakland-Call-Cent/quth-gb8e) from the City of Oakland's Open Data Portal. 

In [2]:
# url = 'https://data.oaklandca.gov/api/views/quth-gb8e/rows.csv?accessType=DOWNLOAD'
# r = requests.get(url, allow_redirects=False)

# # write the content of the request into a file called `oakland_311.csv`
# open('oakland_311.csv', 'wb').write(r.content)

## Import the csv into a `pandas` dataframe

We use a method called `pd.read_csv()` to import a csv file into a dataframe.

Make sure you assign the dataframe into a variable. Below, we're calling the dataframe `oakland_311_original`.

In [3]:
oakland_311_original = pd.read_csv('oakland_311.csv')

This variable `oakland_311_original` will help us download the data so that we can work on it through the pandas dataframe.

## View data

### `df.head()` and `df.tail()`

Use `df.head()` to view the first 5 rows and the last 5 rows of the dataframe. 

In [4]:
oakland_311_original.head()

REQUESTID            DATETIMEINIT       SOURCE  \
0    1167791  12/01/2021 08:29:53 AM        Email   
1    1167997  12/01/2021 02:54:36 PM    Voicemail   
2    1148022  09/16/2021 09:37:36 PM  SeeClickFix   
3    1148574  09/20/2021 01:30:43 PM  SeeClickFix   
4    1149273  09/22/2021 01:20:59 PM  SeeClickFix   

                   DESCRIPTION REQCATEGORY  \
0  Inspections - Sewer Lateral      SEWERS   
1          Homeless Encampment    HE_CLEAN   
2  Parking - Abandoned Vehicle      POLICE   
3  Parking - Abandoned Vehicle      POLICE   
4  Parking - Abandoned Vehicle      POLICE   

                                  REQADDRESS  STATUS REFERREDTO  \
0                                        NaN    OPEN        NaN   
1                                        NaN    OPEN        NaN   
2   (37.84456257751645, -122.26704955148797)    OPEN        NaN   
3  (37.841942007577984, -122.27307401524136)  CLOSED        NaN   
4     (37.8424970155393, -122.2694559662867)    OPEN        NaN   

           DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
0                     NaN          NaN          NaN             NaN  NaN   
1                     NaN          NaN          NaN             NaN  NaN   
2                     NaN  6051460.077  2134803.248            CCD1  11X   
3  12/01/2021 02:06:40 PM  6049702.742  2133882.102            CCD1  10Y   
4                     NaN  6050751.127  2134064.386            CCD1  11X   

                      PROBADDRESS     City State  
0                   2122 108TH AV  Oakland    CA  
1                   OAK GLEN PARK  Oakland    CA  
2                 5911 MACCALL ST  Oakland    CA  
3                     878 57TH ST  Oakland    CA  
4  5714 MARTIN LUTHER KING JR WAY  Oakland    CA

Here are the first 5 rows of our dataset, right now it looks really incomplete we have missing dates and it looks really hard to read, later in pandas, we will learn how to clean our dataset

If you notice to your left, each row is assigned a number (i.e. 0, 1, 2, 3, 4), these are 'index' numbers that help identify each row, in other words a 'row id'. We can use 'index numbers' to call a certain range of rows.

You can use df.head(20) to view the first 20 rows. 

In order to view the last 5 rows of a dataset, we can use df.tail()

In [5]:
oakland_311_original.tail()

REQUESTID            DATETIMEINIT       SOURCE  \
862926    1164947  11/17/2021 08:29:12 AM        Phone   
862927    1165400  11/18/2021 03:04:00 PM        Email   
862928    1167199  11/29/2021 09:04:08 AM        Email   
862929    1164502  11/15/2021 04:52:09 PM  SeeClickFix   
862930    1166301  11/23/2021 08:27:12 AM  SeeClickFix   

                                        DESCRIPTION REQCATEGORY  \
862926           Traffic - Repair Pavement Striping     TRAFFIC   
862927        City Bldg - Electrical Inside/Outside   BLDGMAINT   
862928  Spillage - Trash, Recycle, Compost Spillage   RECYCLING   
862929   Illegal Dumping - debris, appliances, etc.     ILLDUMP   
862930                        Parking - Enforcement     PARKING   

                                       REQADDRESS    STATUS REFERREDTO  \
862926   (37.74587673008652, -122.15610217415544)  WOCREATE        NaN   
862927   (37.80754717953511, -122.27516788700873)    CLOSED        NaN   
862928  (37.794757314318225, -122.19210995683491)      OPEN        NaN   
862929   (37.76940868428093, -122.18206644059013)    CLOSED        NaN   
862930   (37.79240804533573, -122.19340337212121)    CLOSED        NaN   

                DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
862926                     NaN  6082854.999  2098289.251            CCD7  32Y   
862927  11/23/2021 02:21:12 PM  6048860.675  2121371.814            CCD3  04X   
862928                     NaN  6072767.360  2116271.016            CCD4  25X   
862929  11/27/2021 03:26:44 PM  6075503.020  2106990.208            CCD6  30X   
862930  11/23/2021 01:51:00 PM  6072378.218  2115422.494            CCD4  25X   

                           PROBADDRESS     City State  
862926       100TH AV & MACARTHUR BLVD  Oakland    CA  
862927  1601 MARTIN LUTHER KING JR WAY  Oakland    CA  
862928              4063 PATTERSON AVE  Oakland    CA  
862929               6643 BANCROFT AVE  Oakland    CA  
862930                    3911 BAYO ST  Oakland    CA

If you call the dataframe itself, you will get the first 5 rows and the last 5 rows

In [6]:
oakland_311_original

REQUESTID            DATETIMEINIT       SOURCE  \
0         1167791  12/01/2021 08:29:53 AM        Email   
1         1167997  12/01/2021 02:54:36 PM    Voicemail   
2         1148022  09/16/2021 09:37:36 PM  SeeClickFix   
3         1148574  09/20/2021 01:30:43 PM  SeeClickFix   
4         1149273  09/22/2021 01:20:59 PM  SeeClickFix   
...           ...                     ...          ...   
862926    1164947  11/17/2021 08:29:12 AM        Phone   
862927    1165400  11/18/2021 03:04:00 PM        Email   
862928    1167199  11/29/2021 09:04:08 AM        Email   
862929    1164502  11/15/2021 04:52:09 PM  SeeClickFix   
862930    1166301  11/23/2021 08:27:12 AM  SeeClickFix   

                                        DESCRIPTION REQCATEGORY  \
0                       Inspections - Sewer Lateral      SEWERS   
1                               Homeless Encampment    HE_CLEAN   
2                       Parking - Abandoned Vehicle      POLICE   
3                       Parking - Abandoned Vehicle      POLICE   
4                       Parking - Abandoned Vehicle      POLICE   
...                                             ...         ...   
862926           Traffic - Repair Pavement Striping     TRAFFIC   
862927        City Bldg - Electrical Inside/Outside   BLDGMAINT   
862928  Spillage - Trash, Recycle, Compost Spillage   RECYCLING   
862929   Illegal Dumping - debris, appliances, etc.     ILLDUMP   
862930                        Parking - Enforcement     PARKING   

                                       REQADDRESS    STATUS REFERREDTO  \
0                                             NaN      OPEN        NaN   
1                                             NaN      OPEN        NaN   
2        (37.84456257751645, -122.26704955148797)      OPEN        NaN   
3       (37.841942007577984, -122.27307401524136)    CLOSED        NaN   
4          (37.8424970155393, -122.2694559662867)      OPEN        NaN   
...                                           ...       ...        ...   
862926   (37.74587673008652, -122.15610217415544)  WOCREATE        NaN   
862927   (37.80754717953511, -122.27516788700873)    CLOSED        NaN   
862928  (37.794757314318225, -122.19210995683491)      OPEN        NaN   
862929   (37.76940868428093, -122.18206644059013)    CLOSED        NaN   
862930   (37.79240804533573, -122.19340337212121)    CLOSED        NaN   

                DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
0                          NaN          NaN          NaN             NaN  NaN   
1                          NaN          NaN          NaN             NaN  NaN   
2                          NaN  6051460.077  2134803.248            CCD1  11X   
3       12/01/2021 02:06:40 PM  6049702.742  2133882.102            CCD1  10Y   
4                          NaN  6050751.127  2134064.386            CCD1  11X   
...                        ...          ...          ...             ...  ...   
862926                     NaN  6082854.999  2098289.251            CCD7  32Y   
862927  11/23/2021 02:21:12 PM  6048860.675  2121371.814            CCD3  04X   
862928                     NaN  6072767.360  2116271.016            CCD4  25X   
862929  11/27/2021 03:26:44 PM  6075503.020  2106990.208            CCD6  30X   
862930  11/23/2021 01:51:00 PM  6072378.218  2115422.494            CCD4  25X   

                           PROBADDRESS     City State  
0                        2122 108TH AV  Oakland    CA  
1                        OAK GLEN PARK  Oakland    CA  
2                      5911 MACCALL ST  Oakland    CA  
3                          878 57TH ST  Oakland    CA  
4       5714 MARTIN LUTHER KING JR WAY  Oakland    CA  
...                                ...      ...   ...  
862926       100TH AV & MACARTHUR BLVD  Oakland    CA  
862927  1601 MARTIN LUTHER KING JR WAY  Oakland    CA  
862928              4063 PATTERSON AVE  Oakland    CA  
862929               6643 BANCROFT AVE  Oakland    CA  
862930                    39

This is interesting because it looks like Case ID 862928 is still open even though it's one of the older cases. I'm also interested in what "WOCREATE" means, on first glance, it looks like a case that will be created. Hopefully, we can follow this later.

### `df.info()`

Use `df.info()` to get more information on the dataframe. 

In [7]:
oakland_311_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 862931 entries, 0 to 862930
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   REQUESTID        862931 non-null  int64  
 1   DATETIMEINIT     862931 non-null  object 
 2   SOURCE           862931 non-null  object 
 3   DESCRIPTION      862931 non-null  object 
 4   REQCATEGORY      862849 non-null  object 
 5   REQADDRESS       846829 non-null  object 
 6   STATUS           862929 non-null  object 
 7   REFERREDTO       71107 non-null   object 
 8   DATETIMECLOSED   622089 non-null  object 
 9   SRX              846829 non-null  float64
 10  SRY              846829 non-null  float64
 11  COUNCILDISTRICT  819703 non-null  object 
 12  BEAT             822301 non-null  object 
 13  PROBADDRESS      862452 non-null  object 
 14  City             862931 non-null  object 
 15  State            862931 non-null  object 
dtypes: float64(2), int64(1), object(13)
me

So these are all the columns featured in the dataframe. If you notice under "count" each column is represented as "non-null" meaning the column cannot have a null value (or no value). 

### What is a d-type? 

You remember when we talked about Python **types**, like `int`, `float`, and `string`? Above, we have `int64` instead of `int`, and `float64` instead of `float`. (`object` is pretty close to `string`, but not exactly.)

Here, **dtype** stands for **data type** and comes from a module called `numpy`. Even though we did not `import numpy`, the pandas module imported numpy within its own code.

Side note on dtypes: Sometimes, data doesn't import correctly, and you have to `df.read_csv()` again while simultaneously specifying the dtype. We're not going to do that today because it looks like most of the data imported OK. But we will convert the dtype of two columns so we can perform certain calculations.

Let's take a look at a bit of the dataframe again.

In [8]:
oakland_311_original.head()

REQUESTID            DATETIMEINIT       SOURCE  \
0    1167791  12/01/2021 08:29:53 AM        Email   
1    1167997  12/01/2021 02:54:36 PM    Voicemail   
2    1148022  09/16/2021 09:37:36 PM  SeeClickFix   
3    1148574  09/20/2021 01:30:43 PM  SeeClickFix   
4    1149273  09/22/2021 01:20:59 PM  SeeClickFix   

                   DESCRIPTION REQCATEGORY  \
0  Inspections - Sewer Lateral      SEWERS   
1          Homeless Encampment    HE_CLEAN   
2  Parking - Abandoned Vehicle      POLICE   
3  Parking - Abandoned Vehicle      POLICE   
4  Parking - Abandoned Vehicle      POLICE   

                                  REQADDRESS  STATUS REFERREDTO  \
0                                        NaN    OPEN        NaN   
1                                        NaN    OPEN        NaN   
2   (37.84456257751645, -122.26704955148797)    OPEN        NaN   
3  (37.841942007577984, -122.27307401524136)  CLOSED        NaN   
4     (37.8424970155393, -122.2694559662867)    OPEN        NaN   

           DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
0                     NaN          NaN          NaN             NaN  NaN   
1                     NaN          NaN          NaN             NaN  NaN   
2                     NaN  6051460.077  2134803.248            CCD1  11X   
3  12/01/2021 02:06:40 PM  6049702.742  2133882.102            CCD1  10Y   
4                     NaN  6050751.127  2134064.386            CCD1  11X   

                      PROBADDRESS     City State  
0                   2122 108TH AV  Oakland    CA  
1                   OAK GLEN PARK  Oakland    CA  
2                 5911 MACCALL ST  Oakland    CA  
3                     878 57TH ST  Oakland    CA  
4  5714 MARTIN LUTHER KING JR WAY  Oakland    CA

There are 2 columns that we want to convert. **DATETIMEINIT** and **DATETIMECLOSED** are both `object` dtype, but we want to change them to a `datetime64` dtype. That allows us to do some math operations, like sort by the earliest date in the dataframe. This will help us organize the dataframe.

In [9]:
oakland_311_original['DATETIMEINIT'].min()

'01/01/2011 05:51:45 PM'

In [10]:
oakland_311_original['DATETIMEINIT'].max()

'12/31/2020 12:59:06 PM'

The operations above are wrong — they are sorted by string, not by actual date! We saw 2021 data when we called the df.tail() method earlier.  

This is an error in the dataframe that we hope to fix because the 2021 data is not included and those are the most recent cases.

## Properly type the data

### Copy the original dataframe

Before we start convert the 2 columns, let's copy the original dataframe into a new dataframe. Below, I'm going to use `df.copy()` to create a copy of the original dataframe. We're not going to alter the original dataframe at all. That way, if we run into any problems later, we can compare our edited dataframe with the original dataframe.

I realize that this function will be really helpful for us to help figure out the inconsistencies in the dataframe.

In [11]:
oakland_311 = oakland_311_original.copy()

### Convert columns to `datetime`

Let's convert **DATETIMEINIT** first so we can compare the two columns. The formula for this conversion is:

```python
df['column_name'] = pd.to_datetime(df['column_name'])
```

In [12]:
oakland_311['DATETIMEINIT'] = pd.to_datetime(oakland_311['DATETIMEINIT'])

Now, lets look at the dataframe and compare "DATETIMEINIT" to "DATETIMECLOSED"

In [13]:
oakland_311.head()

REQUESTID        DATETIMEINIT       SOURCE                  DESCRIPTION  \
0    1167791 2021-12-01 08:29:53        Email  Inspections - Sewer Lateral   
1    1167997 2021-12-01 14:54:36    Voicemail          Homeless Encampment   
2    1148022 2021-09-16 21:37:36  SeeClickFix  Parking - Abandoned Vehicle   
3    1148574 2021-09-20 13:30:43  SeeClickFix  Parking - Abandoned Vehicle   
4    1149273 2021-09-22 13:20:59  SeeClickFix  Parking - Abandoned Vehicle   

  REQCATEGORY                                 REQADDRESS  STATUS REFERREDTO  \
0      SEWERS                                        NaN    OPEN        NaN   
1    HE_CLEAN                                        NaN    OPEN        NaN   
2      POLICE   (37.84456257751645, -122.26704955148797)    OPEN        NaN   
3      POLICE  (37.841942007577984, -122.27307401524136)  CLOSED        NaN   
4      POLICE     (37.8424970155393, -122.2694559662867)    OPEN        NaN   

           DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
0                     NaN          NaN          NaN             NaN  NaN   
1                     NaN          NaN          NaN             NaN  NaN   
2                     NaN  6051460.077  2134803.248            CCD1  11X   
3  12/01/2021 02:06:40 PM  6049702.742  2133882.102            CCD1  10Y   
4                     NaN  6050751.127  2134064.386            CCD1  11X   

                      PROBADDRESS     City State  
0                   2122 108TH AV  Oakland    CA  
1                   OAK GLEN PARK  Oakland    CA  
2                 5911 MACCALL ST  Oakland    CA  
3                     878 57TH ST  Oakland    CA  
4  5714 MARTIN LUTHER KING JR WAY  Oakland    CA

Can you see the difference between the previous dataframe? The cases are now in descending order. So now let's call df.info() again.

In [14]:
oakland_311.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 862931 entries, 0 to 862930
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   REQUESTID        862931 non-null  int64         
 1   DATETIMEINIT     862931 non-null  datetime64[ns]
 2   SOURCE           862931 non-null  object        
 3   DESCRIPTION      862931 non-null  object        
 4   REQCATEGORY      862849 non-null  object        
 5   REQADDRESS       846829 non-null  object        
 6   STATUS           862929 non-null  object        
 7   REFERREDTO       71107 non-null   object        
 8   DATETIMECLOSED   622089 non-null  object        
 9   SRX              846829 non-null  float64       
 10  SRY              846829 non-null  float64       
 11  COUNCILDISTRICT  819703 non-null  object        
 12  BEAT             822301 non-null  object        
 13  PROBADDRESS      862452 non-null  object        
 14  City             862

You can see that in the "DATETIMEINIT" column, the dtype is now changed from "object" to "datetime64[ns]"

In [15]:
oakland_311['DATETIMECLOSED'] = pd.to_datetime(oakland_311['DATETIMECLOSED'])

Let's now convert the second column "DATETIMECLOSED"

And let's take a peek at the change:

In [16]:
oakland_311.head()

REQUESTID        DATETIMEINIT       SOURCE                  DESCRIPTION  \
0    1167791 2021-12-01 08:29:53        Email  Inspections - Sewer Lateral   
1    1167997 2021-12-01 14:54:36    Voicemail          Homeless Encampment   
2    1148022 2021-09-16 21:37:36  SeeClickFix  Parking - Abandoned Vehicle   
3    1148574 2021-09-20 13:30:43  SeeClickFix  Parking - Abandoned Vehicle   
4    1149273 2021-09-22 13:20:59  SeeClickFix  Parking - Abandoned Vehicle   

  REQCATEGORY                                 REQADDRESS  STATUS REFERREDTO  \
0      SEWERS                                        NaN    OPEN        NaN   
1    HE_CLEAN                                        NaN    OPEN        NaN   
2      POLICE   (37.84456257751645, -122.26704955148797)    OPEN        NaN   
3      POLICE  (37.841942007577984, -122.27307401524136)  CLOSED        NaN   
4      POLICE     (37.8424970155393, -122.2694559662867)    OPEN        NaN   

       DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
0                 NaT          NaN          NaN             NaN  NaN   
1                 NaT          NaN          NaN             NaN  NaN   
2                 NaT  6051460.077  2134803.248            CCD1  11X   
3 2021-12-01 14:06:40  6049702.742  2133882.102            CCD1  10Y   
4                 NaT  6050751.127  2134064.386            CCD1  11X   

                      PROBADDRESS     City State  
0                   2122 108TH AV  Oakland    CA  
1                   OAK GLEN PARK  Oakland    CA  
2                 5911 MACCALL ST  Oakland    CA  
3                     878 57TH ST  Oakland    CA  
4  5714 MARTIN LUTHER KING JR WAY  Oakland    CA

By the way, you might notice that we call `df.head()` and `df.tail()` a lot to check our work. That's OK. Later on, you will be able to run more code at once before calling one of those methods, but for now, it's good to check your work often.

### .min() and .max()

Now we can find the earliest date and the latest date of both columns, by performing `series.min()` and `series.max()` on these series. (There's no equivalent of `df`/dataframe for **series**, unfortunately.)

In [17]:
oakland_311['DATETIMECLOSED'].min()

Timestamp('2006-09-08 15:05:19')

In [18]:
oakland_311['DATETIMECLOSED'].max()

Timestamp('2021-12-12 18:40:48')

In [19]:
oakland_311['DATETIMEINIT'].min()

Timestamp('2009-07-01 08:05:36')

In [20]:
oakland_311['DATETIMEINIT'].max()

Timestamp('2021-12-13 02:59:09')

This shows the difference between first closed case and the most recent case that was closed.

### Get the difference of the two date columns
The "Close_Time" variable will help us show the time difference between when a case was closed and when it was initially opened. 

In [21]:
oakland_311['Close_Time'] = oakland_311['DATETIMECLOSED'] - oakland_311['DATETIMEINIT']

We are going to look at the first five cases to determine the difference between 'DATETIMECLOSED' and 'DATETIMEINIT'.

In [22]:
oakland_311.head()

REQUESTID        DATETIMEINIT       SOURCE                  DESCRIPTION  \
0    1167791 2021-12-01 08:29:53        Email  Inspections - Sewer Lateral   
1    1167997 2021-12-01 14:54:36    Voicemail          Homeless Encampment   
2    1148022 2021-09-16 21:37:36  SeeClickFix  Parking - Abandoned Vehicle   
3    1148574 2021-09-20 13:30:43  SeeClickFix  Parking - Abandoned Vehicle   
4    1149273 2021-09-22 13:20:59  SeeClickFix  Parking - Abandoned Vehicle   

  REQCATEGORY                                 REQADDRESS  STATUS REFERREDTO  \
0      SEWERS                                        NaN    OPEN        NaN   
1    HE_CLEAN                                        NaN    OPEN        NaN   
2      POLICE   (37.84456257751645, -122.26704955148797)    OPEN        NaN   
3      POLICE  (37.841942007577984, -122.27307401524136)  CLOSED        NaN   
4      POLICE     (37.8424970155393, -122.2694559662867)    OPEN        NaN   

       DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
0                 NaT          NaN          NaN             NaN  NaN   
1                 NaT          NaN          NaN             NaN  NaN   
2                 NaT  6051460.077  2134803.248            CCD1  11X   
3 2021-12-01 14:06:40  6049702.742  2133882.102            CCD1  10Y   
4                 NaT  6050751.127  2134064.386            CCD1  11X   

                      PROBADDRESS     City State       Close_Time  
0                   2122 108TH AV  Oakland    CA              NaT  
1                   OAK GLEN PARK  Oakland    CA              NaT  
2                 5911 MACCALL ST  Oakland    CA              NaT  
3                     878 57TH ST  Oakland    CA 72 days 00:35:57  
4  5714 MARTIN LUTHER KING JR WAY  Oakland    CA              NaT

## A brief detour into data analysis

Now that we've converted the columns, we can do some interesting operations on them.

### Mean

Get the mean of a column by calling `series.mean()`.

In [23]:
oakland_311['Close_Time'].mean()

Timedelta('50 days 11:09:22.326879275')

The average time the city took to close a case was around 50 days. This is for the whole dataset, from early 2009 to now. *Make sure to calculate the min of 'DATETIMECLOSED' and 'DATETIMEINIT' 

### Median

In [24]:
oakland_311['Close_Time'].median()

Timedelta('4 days 22:26:45')

But the median time was around 4 days.

### Min

In [25]:
oakland_311['Close_Time'].min()

Timedelta('-2981 days +00:37:47')

The shortest amount of time was 0 days and 0 seconds. It could possibly be a public comment that didn't require follow-up. We will get back to this.

## Max

In [26]:
oakland_311['Close_Time'].max()

Timedelta('4293 days 21:59:22')

One case seems to have taken 4293 days! That's almost 11+ years. That seems like way too long. There might be an error here.

Let's take a quick detour into _subsetting_ the data. That means to take a smaller set of the data based on some conditions. I'll explain how to subset more later, but for now, check out the following code:

In [27]:
oakland_311[oakland_311['Close_Time'] >= '4293 days']

REQUESTID        DATETIMEINIT          SOURCE  \
41618     323595 2009-12-09 14:30:28  Phone or Email   

                       DESCRIPTION REQCATEGORY  \
41618  Street Medians (Landscaped)       PARKS   

                                     REQADDRESS  STATUS REFERREDTO  \
41618  (37.82206394742191, -122.20671960118034)  CLOSED        NaN   

           DATETIMECLOSED        SRX        SRY COUNCILDISTRICT BEAT  \
41618 2021-09-11 12:29:50  6068728.0  2126289.0            CCD4  22X   

                     PROBADDRESS     City State         Close_Time  
41618  PARK BLVD & MOUNTAIN BLVD  Oakland    CA 4293 days 21:59:22

In [28]:
oakland_311[oakland_311['Close_Time'] < '0 days']

REQUESTID        DATETIMEINIT          SOURCE  \
4079       432312 2012-11-01 11:24:56  Phone or Email   
5965        23385 2010-09-16 13:59:19  Phone or Email   
6064        23396 2010-09-16 13:59:46  Phone or Email   
6163        24888 2011-04-06 16:33:06  Phone or Email   
6335        25105 2011-05-16 11:55:22  Phone or Email   
...           ...                 ...             ...   
216729     522842 2014-06-10 15:50:45  Phone or Email   
219489     522752 2014-06-10 11:29:36  Phone or Email   
220546     522805 2014-06-10 14:02:16  Phone or Email   
221487     528272 2014-07-10 12:20:22     SeeClickFix   
235866     549531 2014-11-06 14:27:32           Phone   

                                            DESCRIPTION REQCATEGORY  \
4079                                               PG&E       OTHER   
5965               Residential Recycling Service Issues   RECYCLING   
6064               Residential Recycling Service Issues   RECYCLING   
6163    Garbage - Requests Exemption - Weekly or Shared   RECYCLING   
6335               Residential Recycling Service Issues   RECYCLING   
...                                                 ...         ...   
216729                  Recycling - Repair/Replace Cart   RECYCLING   
219489             Recycling - Missed Pickup - Backyard   RECYCLING   
220546                  Recycling - Repair/Replace Cart   RECYCLING   
221487                              Hazardous Materials        FIRE   
235866             Illegal Dumping � mattress/boxspring     ILLDUMP   

                                       REQADDRESS  STATUS  \
4079    (37.790247170707254, -122.23014783691161)  CLOSED   
5965    (30.009899650831787, -141.21875098935422)  CLOSED   
6064    (30.009899650831787, -141.21875098935422)  CLOSED   
6163    (30.009899650831787, -141.21875098935422)  CLOSED   
6335    (30.009899650831787, -141.21875098935422)  CLOSED   
...                                           ...     ...   
216729   (37.81389394379093, -122.25288667849271)  CLOSED   
219489   (37.81947378040759, -122.19864842876396)  CLOSED   
220546    (37.82652163613099, -122.2607234958911)  CLOSED   
221487  (37.823866701293426, -122.26193144921672)  CLOSED   
235866  (37.823866701293426, -122.26193144921672)  CLOSED   

                                  REFERREDTO      DATETIMECLOSED         SRX  \
4079                                   PG&E, 2012-07-10 14:00:14  6061748.00   
5965                                     NaN 2010-09-16 13:59:07        0.00   
6064                                     NaN 2010-09-16 13:59:35        0.00   
6163                                     NaN 2011-04-06 16:33:04        0.00   
6335                                     NaN 2011-05-16 11:55:21        0.00   
...                                      ...                 ...         ...   
216729  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6055339.74   
219489  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6071041.76   
220546  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6053162.82   
221487                 FIRE (GRIFFIN, LEROY) 2006-09-08 15:05:19  6052795.80   
235866                                   NaN 2006-09-08 15:05:19  6052795.80   

               SRY COUNCILDISTRICT BEAT          PROBADDRESS     City State  \
4079    2114830.00             NaN  20X  E 23RD ST & 23RD AV  Oakland    CA   
5965          0.00             NaN  NaN                   ZZ  Oakland    CA   
6064          0.00             NaN  NaN                   ZZ  Oakland    CA   
6163          0.00             NaN  NaN                   ZZ  Oakland    CA   
6335          0.00             NaN  NaN                   ZZ  Oakland    CA   
...            ...             ...  ...                  ...      ...   ...   
216729  2123561.09            CCD3  14X        302 EUCLID AV  Oakland    CA   
219489  2125303.60            CCD4  13Z    2300 MASTLANDS DR  Oakland    CA   
220546  2128200.82            CCD1  08X          409 38TH ST  Oak

It looks like the case that took the longest was an improvement to a local park and the shortest turnaround times are more related to recyclcing and waste issues, which makes sense.

### Sort Data 

We can even look at the top 10 cases that took the longest time to resolve. You'll use the `df.sort_values()` method.

Let's break down the below code before we run it. You can see there are 2 options within the parentheses for `.sort_values()`:
```python
by=['Close_Time'], ascending=False

The `by` argument tells us which column we will sort the dataframe by. You always need to include this argument. You can sort by multiple columns, too.

The optional `ascending` argument tells us if we want the dataframe to sort from smallest to largest or earliest to latest. By default, `ascending` is set to `True`, so we are going to change it here so it is `False`.

Next, I do not want to see the entire dataframe, just the first 10 rows. So I am going to call `df.head(10)`.

In [29]:
oakland_311.sort_values(by=['Close_Time'], ascending=False).head(10)

REQUESTID        DATETIMEINIT          SOURCE  \
41618     323595 2009-12-09 14:30:28  Phone or Email   
55971     356767 2011-01-07 09:57:30         Website   
45852     358846 2011-01-26 14:30:53  Phone or Email   
74657     361419 2011-02-18 10:19:34  Phone or Email   
70807     363868 2011-03-17 10:14:40  Phone or Email   
74085     364257 2011-03-22 07:25:10  Phone or Email   
74835     365654 2011-04-05 10:51:44  Phone or Email   
50138     357227 2011-01-11 16:11:55         Website   
52409     356540 2011-01-05 21:40:17         Website   
80025     367154 2011-04-19 09:22:11  Phone or Email   

                                DESCRIPTION   REQCATEGORY  \
41618           Street Medians (Landscaped)         PARKS   
55971                     Sidewalk - Damage  ROW_STREETSW   
45852                     Sidewalk - Damage  ROW_STREETSW   
74657                     Sidewalk - Damage  ROW_STREETSW   
70807  Streets/Sidewalks - Unsafe Condition  ROW_STREETSW   
74085  Streets/Sidewalks - Unsafe Condition  ROW_STREETSW   
74835  Streets/Sidewalks - Unsafe Condition  ROW_STREETSW   
50138                     Sidewalk - Damage  ROW_STREETSW   
52409  Streets/Sidewalks - Unsafe Condition  ROW_STREETSW   
80025                     Sidewalk - Damage  ROW_STREETSW   

                                      REQADDRESS  STATUS REFERREDTO  \
41618   (37.82206394742191, -122.20671960118034)  CLOSED        NaN   
55971   (37.84714892555996, -122.25195910487346)  CLOSED        NaN   
45852  (37.802893854302425, -122.26918417013395)  CLOSED        NaN   
74657   (37.73308180727985, -122.18820984423338)  CLOSED        NaN   
70807   (37.78232555909755, -122.21188366883626)  CLOSED        NaN   
74085   (37.77778508555348, -122.22563897756443)  CLOSED        NaN   
74835   (37.816837027535065, -122.2568471359077)  CLOSED        NaN   
50138   (37.73561875315405, -122.17677744021242)  CLOSED        NaN   
52409    (37.81062199910106, -122.2589020860214)  CLOSED        NaN   
80025   (37.82987853198286, -122.25723549498862)  CLOSED        NaN   

           DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
41618 2021-09-11 12:29:50  6068728.000  2126289.000            CCD4  22X   
55971 2021-10-29 16:15:09  6055834.314  2135662.921            CCD1  12Y   
45852 2021-10-21 11:28:26  6050557.000  2119645.000            CCD2  03X   
74657 2021-10-29 13:34:38  6073489.000  2093797.000            CCD7  31Y   
70807 2021-10-29 12:24:49  6066972.000  2111849.000            CCD5  24X   
74085 2021-10-28 12:24:59  6062967.000  2110269.000            CCD5  23X   
74835 2021-10-29 15:29:41  6054216.000  2124654.000            CCD3  08X   
50138 2021-08-05 09:29:00  6076811.090  2094661.167            CCD7  31Z   
52409 2021-07-14 09:57:37  6053579.984  2122402.509            CCD3  14X   
80025 2021-10-18 13:03:17  6054193.000  2129404.000            CCD1  12X   

                     PROBADDRESS     City State         Close_Time  
41618  PARK BLVD & MOUNTAIN BLVD  Oakland    CA 4293 days 21:59:22  
55971          5900 COLLEGE AVE.  Oakland    CA 3948 days 06:17:39  
45852            1323 WEBSTER ST  Oakland    CA 3920 days 20:57:33  
74657             310 CASWELL AV  Oakland    CA 3906 days 03:15:04  
70807               2217 38TH AV  Oakland    CA 3879 days 02:10:09  
74085          1418 FRUITVALE AV  Oakland    CA 3873 days 04:59:49  
74835             287 OAKLAND AV  Oakland    CA 3860 days 04:37:57  
50138            10306 PIPPIN ST  Oakland    CA 3858 days 17:17:05  
52409          325 PARK VIEW TER  Oakland    CA 3842 days 12:17:20  
80025             4146 MANILA AV  Oakland    CA 3835 days 03:41:06

It looks like the longest cases are also related to infrastructure. 

## Clean data, part 2: Check and vet the data

### Unique identifier for every row?

First, I want to see if `Case_ID` has a unique ID for every row. Why? When you're doing a data analysis, every row should have its own unique ID. Hopefully, the agency that gave you the data has provided a unique ID. Sometimes, though, they don't. In those cases, you want to create a unique ID for every row.

I get a count of unique values by calling `.nunique()` on a column.

In [30]:
oakland_311['REQUESTID'].nunique()

862931

We can use `df.info()` to get the number of rows, or we can scroll up to see again. To simplify the process, let's just call `len(df)`. We also learned `len()` for both strings and data structures earlier in the semester. 

In [31]:
len(oakland_311)

862931

So there are 862,834 unique case IDs, and 862,834 rows in this dataframe. 

We're going to check by **subsetting** the data.

### Subsetting

This is the general structure of how you subset data in pandas.

```python
df[ expression ]
```

That's not very descriptive. What's the _expression_? There are lots of different ways we write these expressions in pandas. I'm going to show you a handful of different kinds today, but know there are a bunch more!

Ultimately, what I want is a list of the duplicate `Case_ID`s. I'll then subset the dataframe to show any row that has a Case_ID that is on that list.


We'll first check to see which rows in `oakland_311` are exact duplicates.

In [32]:
oakland_311[oakland_311.duplicated()]

Empty DataFrame
Columns: [REQUESTID, DATETIMEINIT, SOURCE, DESCRIPTION, REQCATEGORY, REQADDRESS, STATUS, REFERREDTO, DATETIMECLOSED, SRX, SRY, COUNCILDISTRICT, BEAT, PROBADDRESS, City, State, Close_Time]
Index: []

There are no exact duplicates. Now let's check specifically to see rows in which `Case_ID`s are duplicated. 

In [33]:
oakland_311[oakland_311['REQUESTID'].duplicated()]

Empty DataFrame
Columns: [REQUESTID, DATETIMEINIT, SOURCE, DESCRIPTION, REQCATEGORY, REQADDRESS, STATUS, REFERREDTO, DATETIMECLOSED, SRX, SRY, COUNCILDISTRICT, BEAT, PROBADDRESS, City, State, Close_Time]
Index: []

First, I'm going to create a new dataframe that has the duplicated IDs. We're going to create a copy.

In [34]:
dupe_cases = oakland_311[oakland_311['REQUESTID'].duplicated()].copy()
dupe_cases

Empty DataFrame
Columns: [REQUESTID, DATETIMEINIT, SOURCE, DESCRIPTION, REQCATEGORY, REQADDRESS, STATUS, REFERREDTO, DATETIMECLOSED, SRX, SRY, COUNCILDISTRICT, BEAT, PROBADDRESS, City, State, Close_Time]
Index: []

Now we create a `list` or `array` (the numpy version of a list) of those IDs:

In [35]:
dupe_case_ids = dupe_cases['REQUESTID'].to_list()

# to create an array, you can use this instead:
# dupe_case_ids = dupe_cases['Case_ID'].unique()

Now we'll call dupe_case_ids, so we can see what's in it.

In [36]:
dupe_case_ids

[]

Now we'll subset the data by finding cases in our edited dataframe `oakland_311`.

In [37]:
oakland_311[oakland_311['REQUESTID'].isin(dupe_case_ids)]

Empty DataFrame
Columns: [REQUESTID, DATETIMEINIT, SOURCE, DESCRIPTION, REQCATEGORY, REQADDRESS, STATUS, REFERREDTO, DATETIMECLOSED, SRX, SRY, COUNCILDISTRICT, BEAT, PROBADDRESS, City, State, Close_Time]
Index: []

We still see that there aren't any duplicates in the data. I'm going to sort that dataframe by **REQUESTID**.

In [38]:
oakland_311[oakland_311['REQUESTID'].isin(dupe_case_ids)].sort_values(by=['REQUESTID'])

Empty DataFrame
Columns: [REQUESTID, DATETIMEINIT, SOURCE, DESCRIPTION, REQCATEGORY, REQADDRESS, STATUS, REFERREDTO, DATETIMECLOSED, SRX, SRY, COUNCILDISTRICT, BEAT, PROBADDRESS, City, State, Close_Time]
Index: []

FYI, all the code we used above didn't change the original dataframe. We were subsetting, but we did not subset with a new variable. 

Now, we're going to drop the duplicated cases, and reset the variable `oakland_311`.

In [39]:
oakland_311 = oakland_311.drop_duplicates(subset=['REQUESTID'])

In [40]:
oakland_311

REQUESTID        DATETIMEINIT       SOURCE  \
0         1167791 2021-12-01 08:29:53        Email   
1         1167997 2021-12-01 14:54:36    Voicemail   
2         1148022 2021-09-16 21:37:36  SeeClickFix   
3         1148574 2021-09-20 13:30:43  SeeClickFix   
4         1149273 2021-09-22 13:20:59  SeeClickFix   
...           ...                 ...          ...   
862926    1164947 2021-11-17 08:29:12        Phone   
862927    1165400 2021-11-18 15:04:00        Email   
862928    1167199 2021-11-29 09:04:08        Email   
862929    1164502 2021-11-15 16:52:09  SeeClickFix   
862930    1166301 2021-11-23 08:27:12  SeeClickFix   

                                        DESCRIPTION REQCATEGORY  \
0                       Inspections - Sewer Lateral      SEWERS   
1                               Homeless Encampment    HE_CLEAN   
2                       Parking - Abandoned Vehicle      POLICE   
3                       Parking - Abandoned Vehicle      POLICE   
4                       Parking - Abandoned Vehicle      POLICE   
...                                             ...         ...   
862926           Traffic - Repair Pavement Striping     TRAFFIC   
862927        City Bldg - Electrical Inside/Outside   BLDGMAINT   
862928  Spillage - Trash, Recycle, Compost Spillage   RECYCLING   
862929   Illegal Dumping - debris, appliances, etc.     ILLDUMP   
862930                        Parking - Enforcement     PARKING   

                                       REQADDRESS    STATUS REFERREDTO  \
0                                             NaN      OPEN        NaN   
1                                             NaN      OPEN        NaN   
2        (37.84456257751645, -122.26704955148797)      OPEN        NaN   
3       (37.841942007577984, -122.27307401524136)    CLOSED        NaN   
4          (37.8424970155393, -122.2694559662867)      OPEN        NaN   
...                                           ...       ...        ...   
862926   (37.74587673008652, -122.15610217415544)  WOCREATE        NaN   
862927   (37.80754717953511, -122.27516788700873)    CLOSED        NaN   
862928  (37.794757314318225, -122.19210995683491)      OPEN        NaN   
862929   (37.76940868428093, -122.18206644059013)    CLOSED        NaN   
862930   (37.79240804533573, -122.19340337212121)    CLOSED        NaN   

            DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
0                      NaT          NaN          NaN             NaN  NaN   
1                      NaT          NaN          NaN             NaN  NaN   
2                      NaT  6051460.077  2134803.248            CCD1  11X   
3      2021-12-01 14:06:40  6049702.742  2133882.102            CCD1  10Y   
4                      NaT  6050751.127  2134064.386            CCD1  11X   
...                    ...          ...          ...             ...  ...   
862926                 NaT  6082854.999  2098289.251            CCD7  32Y   
862927 2021-11-23 14:21:12  6048860.675  2121371.814            CCD3  04X   
862928                 NaT  6072767.360  2116271.016            CCD4  25X   
862929 2021-11-27 15:26:44  6075503.020  2106990.208            CCD6  30X   
862930 2021-11-23 13:51:00  6072378.218  2115422.494            CCD4  25X   

                           PROBADDRESS     City State       Close_Time  
0                        2122 108TH AV  Oakland    CA              NaT  
1                        OAK GLEN PARK  Oakland    CA              NaT  
2                      5911 MACCALL ST  Oakland    CA              NaT  
3                          878 57TH ST  Oakland    CA 72 days 00:35:57  
4       5714 MARTIN LUTHER KING JR WAY  Oakland    CA              NaT  
...                                ...      ...   ...              ...  
862926       100TH AV & MACARTHUR BLVD  Oakland    CA              NaT  
862927  1601 MARTIN LUTHER KING JR WAY  Oakland    CA  4 days 23:17:12  
862928              4063 PATTERSON AVE  Oakland    CA              NaT  
862929    

By the way, you might have noticed there's something called `NaT` in one of the rows above. `NaT` stands for _not a time_ and is kind of like `None` or an empty cell in Google Sheets. For non-time related blank cells, you'll see `NaN` (not a number) instead of `NaT`.

The difference between `None` and `NaN`/`NaT` is that the latter allows you to perform calcuations and skip any blank cells. That means, you probably need to check for how many `NaN`/`NaT` cells exist in your dataframe. If there are a lot of them, your analysis might not be valid. You can quickly check for that with the `df.info()` method we learned earlier. There's a column called `Non-Null Count`. 

In [41]:
oakland_311.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 862931 entries, 0 to 862930
Data columns (total 17 columns):
 #   Column           Non-Null Count   Dtype          
---  ------           --------------   -----          
 0   REQUESTID        862931 non-null  int64          
 1   DATETIMEINIT     862931 non-null  datetime64[ns] 
 2   SOURCE           862931 non-null  object         
 3   DESCRIPTION      862931 non-null  object         
 4   REQCATEGORY      862849 non-null  object         
 5   REQADDRESS       846829 non-null  object         
 6   STATUS           862929 non-null  object         
 7   REFERREDTO       71107 non-null   object         
 8   DATETIMECLOSED   622089 non-null  datetime64[ns] 
 9   SRX              846829 non-null  float64        
 10  SRY              846829 non-null  float64        
 11  COUNCILDISTRICT  819703 non-null  object         
 12  BEAT             822301 non-null  object         
 13  PROBADDRESS      862452 non-null  object         
 14  City

Which columns have a lot of null values? I'm guessing that column with the least non-null values have more null values, and it looks like 'DATETIMECLOSED' has the least non-null values, which makes sense because many of these cases are still in progress and haven't been closed yet.

### Assert
The keyword `assert` is a good way for us to check if the length of the dataframe now matches the number of unique IDs.

In [42]:
assert len(oakland_311) == oakland_311['REQUESTID'].nunique()

If the assertion is `True`, nothing happens. But if the assertion is `False`, you'll get an error. You might want to use these kinds of assertions when you have to re-run your notebooks or have to import updated datasets.

## Export a clean version of the data to a csv

We can use the `df.to_csv()` method to export a clean copy of the csv. That way, you can instantly import the clean data in a new notebook instead of rerunning the code in this notebook.

Before we run the code below, let's take a closer look:

```python
berkeley_311.to_csv('berkeley_311_clean.csv', index=False)
```

The first argument in `df.to_csv()` is the name of the file we're going to export our dataframe into. In this case, that's `berkeley_311_clean.csv`.

The second argument is `index=False`. This means that I don't want pandas to export those row ids (0, 1, 2, 3, etc.) that show up at the very lefthand side of the dataframe.

In [48]:
oakland_311.to_csv('oakland_311_clean.csv', index=False)

Try removing `index=False` and see what happens.

When I tried removing `index=False` there was no change to the dataframe, I'm not sure why but I'm guessing that the `oakland_311_clean.csv` automatically removes the index numbers when transferred to a `csv`.